File: WitchcraftApp/src/main/java/com/tarbadev/witchcraft/recipes/persistence/entity/StepEntity.java
package com.tarbadev.witchcraft.recipes.persistence.entity;

import com.tarbadev.witchcraft.recipes.domain.entity.Step;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Entity
public class StepEntity {
  @Id
  @GeneratedValue(strategy= GenerationType.AUTO)
  private Integer id;
  @Column(length = 1000)
  private String name;

  public StepEntity(Step step) {
    id = step.getId();
    name = step.getName();
  }

  public Step step() {
    return Step.builder()
        .id(id)
        .name(name)
        .build();
  }
}

File: WitchcraftApp/src/main/kotlin/com/tarbadev/witchcraft/recipes/persistence/entity/StepEntity.kt
package com.tarbadev.witchcraft.recipes.persistence.entity

import com.tarbadev.witchcraft.recipes.domain.entity.Step
import com.tarbadev.witchcraft.recipes.domain.entity.StepNote
import javax.persistence.*

@Entity(name = "step")
data class StepEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Int = 0,

    @Column(length = 1000)
    val name: String = "",

    @OneToMany(cascade = [CascadeType.ALL])
    @JoinColumn(name = "step_id")
    val notes: List<StepNoteEntity> = emptyList()
) {

  constructor(step: Step) : this(
      id = step.id,
      name = step.name,
      notes = if (step.note.comment.isNotEmpty())
        listOf(StepNoteEntity(step.note))
      else
        emptyList()
  )

  fun toStep(): Step {
    return Step(
        id = id,
        name = name,
        note = notes.map { it.toStepNote() }.firstOrNull() ?: StepNote()
    )
  }
}
